src/rviz/ogre_helpers/stl_loader.cpp

```diff
@@ -82,15 +84,50 @@ bool STLLoader::load(const std::string& path)
   }
   fclose( input );
 
-  bool success = load(buffer);
-  delete [] buffer;
+  return this->load(buffer, num_bytes_read, path);
+}
 
-  return success;
+bool STLLoader::load(uint8_t* buffer, const size_t num_bytes, const std::string& origin)
+{
+  // check for ascii since we can only load binary types with this class
+  std::string buffer_str = std::string((char*)buffer,num_bytes);
+
+  if(buffer_str.substr(0,5)  == std::string("solid")) 
+  {
+    // file says that it is ascii, but why should we trust it?
+
+    // check for "endsolid" as well
+    if(buffer_str.find("endsolid",5) != std::string::npos) {
+      ROS_ERROR_STREAM("Your STL file \""<<origin<<"\" is malformed. It starts with the word \"solid\" and also contains the word \"endsolid\", indicating that it's an ASCII STL file, but RVIZ can only load binary STL files so it will not be loaded. Please convert the file to a binary format.");
+      return false;
+    }
+
+    // chastise the user for malformed files
+    ROS_WARN_STREAM("Your STL file \""<<origin<<"\" is malformed. It starts with the word \"solid\", indicating that it's an ASCII STL file, but it does not contain the word \"endsolid\" so it is either a malformed ASCII STL file or it is actually a binary STL file. Proceeding assuming binary format.");
+  }
+
+  // make sure there's enough data for a binary STL header and triangle count
+  if(num_bytes <= 84) {
+    ROS_ERROR_STREAM("Your STL file \""<<origin<<"\" is malformed. It appears to be a binary file but does not contain enough data for the 80 byte header and 16-bit integer triangle count");
+    return false;
+  }
+
+  // we're pretty sure this is a malformed binary file at this point, one last check to make sure it won't segfault
```

Comment:
This comment claims that this is a malformed binary file at this point. I think we cannot assume this here, all the checks are good so far.